### ISSUE TITLE
traffic splitting with session affinity

### ISSUE THREAD

**Issue by vadimeisenbergibm** [envoyproxy/envoy#8167]
Can Envoy perform traffic splitting while preserving session affinity? For example, use Ring Hash or Maglev to select the weighted cluster according to its weight? This way, when `route.RouteAction.HashPolicy` is specified, the same weighted cluster and the same endpoint point will be used by applying Ring Hash/Maglev twice, for selecting the weighted cluster and for selecting the endpoint.

**Comment by zuercher** [comment]
This isn't currently possible. Weighted clusters are chosen based on a random value selected at request time. I'll go ahead and label this as an enhancement request.

**Comment by vadimeisenbergibm** [comment]
@zuercher Thanks!

**Comment by zuercher** [comment]
That seems like a reasonable feature to add.

**Comment by temporafugiunt** [comment]
@zuercher Thank you, that is a good idea and I will try that!

> That doesn't sound like a weighted cluster routing problem. It sounds like either the upstream server itself is truncating the data or something about how you're updating Envoy's config is causing a problem. I would look at Envoy's log to see what's going on.

Sorry for the n00b question... But can you recommend any good documentation resources on how to debug envoy to determine root cause for an issue like this? If not I will have a discussion with Dr. Google for a while to see if I can find something.

**Comment by zuercher** [comment]
I would enable debug logging (`-l debug`). You should be able to see the request and response headers for both the downstream and upstream requests. Beyond that, you might have better luck asking in the Envoy slack channel (with logs & config snippets) or else opening a separate issue.

**Comment by vietwow** [comment]
Hi, any news for this feature ?

**Comment by Jeskz0rd** [comment]
Any suggestions for a workaround? :thinking:

**Comment by petrukngantuk** [comment]
any update about this?
I try to use websocket with weight routing in istio, but blocked by this... to enable flagger canary on my deployment
:sob:

**Comment by rgs1** [comment]
> Can Envoy perform traffic splitting while preserving session affinity? For example, use Ring Hash or Maglev to select the weighted cluster according to its weight? This way, when `route.RouteAction.HashPolicy` is specified, the same weighted cluster and the same endpoint point will be used by applying Ring Hash/Maglev twice, for selecting the weighted cluster and for selecting the endpoint.

We actually have an internal filter that does something like that... More info here:

https://medium.com/pinterest-engineering/simplifying-web-deploys-19244fe13737

If there's interest, we could chat a bit more and I might be able to get some cycles to upstream our filter.

**Comment by vadimeisenbergibm** [comment]
@rgs1 Thanks, I switched to work on another domain, but maybe other folks who are interested in this feature, could proceed with this issue.

**Comment by rafaeldasilva** [comment]
By I could understand the point is apply weight rule only on first call.
Since the client connected the next calls goes through the stick rule.
Is it feasible?
Any one see any drawback on this logic?

**Comment by petrukngantuk** [comment]
> By I could understand the point is apply weight rule only on first call.
> Since the client connected the next calls goes through the stick rule.
> 
> Would be an “IF” before apply weight. Where if there is a sticky match the call bypass the weight rule and goes to destination.
> Is it feasible?
> Any one see any drawback on this logic?

personally not, but not sure from maintainer / other

btw, why don't we detect it through Header Upgrade and or Connection Upgrade?

**Comment by benpoulson** [comment]
Nginx currently splits by creating a MurmurHash2 number (0-4294967295) based on specific variables (eg remote IP)

See: http://nginx.org/en/docs/http/ngx_http_split_clients_module.html

This allows users to be consistently sent to the same upstream.

**Comment by defool** [comment]
Any update for this feature?

**Comment by sschepens** [comment]
Maybe https://github.com/envoyproxy/envoy/pull/14875 will fix this.

**Comment by arungeorge101** [comment]
Is this feature being worked on and when will it be available?

**Comment by jcrugzz** [comment]
@rgs1 Would definitely be interesting in chatting sometime if you wanted.

**Comment by rgs1** [comment]
> @rgs1 Would definitely be interesting in chatting sometime if you wanted.

Sure ping me on Slack and we can coordinate, I actually gave some extra thought this week on what it'd take to open source what's described in the above blog post... But couldn't allocate the cycles yet.

**Comment by howardjohn** [comment]
@snowp does https://github.com/envoyproxy/envoy/pull/14875 solve this case or is it not applicable?

**Comment by snowp** [comment]
That PR adds support for a consistent matcher to the new matching framework, but that hasn't yet been wired up to work with the router to allow its usage to construct the route table (and I don't think anyone has planned this work yet).

You could probably get something working by using a simple filter that uses the consistent matcher to inject a cluster header that can then inform the the router, or even using the new `setRoute` API to influence the route selection directly without having to work through the relatively large effort to integrate the new matching framework with the router.

**Comment by mksha** [comment]
Any updates on it ?

**Comment by mksha** [comment]
@vadimeisenbergibm did you find any workaround it ?

**Comment by vadimeisenbergibm** [comment]
@mksha No, I did not, sorry.

**Comment by mksha** [comment]
@zuercher are we planning to add this feature in near future?

**Comment by timvan** [comment]
Hi, just commenting to show interest for this feature. The use case would be with experimentation. Traffic should be split between control and variant models/apps and users must be directed to a consistent variant. Bucketing is based on a user id header.

The only workaround i've seen so far is this:
https://github.com/istio/istio/issues/9764#issuecomment-631994703

**Comment by gongzh** [comment]
> Hi, just commenting to show interest for this feature. The use case would be with experimentation. Traffic should be split between control and variant models/apps and users must be directed to a consistent variant. Bucketing is based on a user id header.

Hello, is there any progress or plan on this feature? thanks!

**Comment by mksha** [comment]
is there a plan to work on it, bec its very generic usecase for everyone.

**Comment by alyssawilk** [comment]
The Envoy project doesn't have paid developers, it is an open source community where anyone can contribute.  So largely features don't happen until someone wants them enough to implement them or pay someone like Tetrate to do so.

**Comment by mksha** [comment]
@rootsongjc Do we have this feature in upstream now?

**Comment by rootsongjc** [comment]
@mksha I don't think so.

**Comment by jcometki** [comment]
I had this same issue, I solved that with the following configuration:

```yaml
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: sample-virtual-service
spec:
  gateways:
    - default
  hosts:
    - sample-host
  http:
    - match:
        - headers:
            Cookie:
              regex: .*userAffinity=v1.*
      route:
        - destination:
            host: sample-host.svc.cluster.local
            port:
              number: 80
            subset: v1
    - match:
        - headers:
            Cookie:
              regex: .*userAffinity=v2.*
      route:
        - destination:
            host: sample-host.svc.cluster.local
            port:
              number: 80
            subset: v2
    - route:
        - destination:
            host: sample-host.svc.cluster.local
            port:
              number: 80
            subset: v1
          weight: 75
        - destination:
            host: sample-host.svc.cluster.local
            port:
              number: 80
            subset: v2
          headers:
            response:
              add:
                Set-Cookie: userAffinity=v2; Max-Age=3600000
          weight: 25
---
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: sample-destination-rule
spec:
  host: sample-host.svc.cluster.local
  subsets:
    - labels:
        app: sample-app-v1
      name: v1
      trafficPolicy:
        loadBalancer:
          consistentHash:
            httpCookie:
              name: userAffinity=v1
              ttl: 1h
    - labels:
        app: sample-app-v2
      name: v2
```

**Comment by mksha** [comment]
@jcometki We have something similar in place, but i have a question about DestinationRule that you have, May I know what is the need of having `name: userAffinity=v1` rather than just `name: userAffinity`
Ideally name property is to represent name of the cookie but the name:value pair. May be I am missing something, but would love to understand your point as well.

**Comment by mksha** [comment]
https://istio.io/latest/docs/reference/config/networking/destination-rule/#LoadBalancerSettings-ConsistentHashLB-HTTPCookie

**Comment by itnazeer** [comment]
Did any one got this working?

**Comment by CaptainGopher** [comment]
Any updates or workaround? : )

**Comment by zuercher** [comment]
I'm going to close this ticket. Many of the recent comments have been for projects that build on Envoy and as such aren't really appropriate here. 

See https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html#config-route-v3-routeaction-hashpolicy-cookie and https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/stateful_session_filter#config-http-filters-stateful-session for various types of session affinity that are currently supported. The former implements the feature originally requested.

**Comment by rixongary** [comment]
@jcometki hello. thanks for the config you posted above. I know this was 2 years ago now, but I am wondering if you could provide any more information on how this worked out for you?

I think there would be a problem where once a user hits the v2 subset and gets the cookie, all of their subsequent requests would be sent to v2, but *those requests would not be taken into account for weighting calculations*, therefore, you would end up with more than 5% of requests going to the v2 subset, and this problem would get worse and worse the higher the canary weight was set.

If anyone has a solution for this, I'd love to hear. Thanks!
